test/services/tsheets_service_test.rb

```diff
@@ -1,26 +1,27 @@
 require 'test_helper'
 
-class TsheetsServiceTest < ActiveSupport::TestCase
-  def setup
-    @service = TsheetsService.new
-  end
-
-  def stub_users
+class TsheetsAdapterFake
+  def find_users
     valid_user = TSheets::Models::User.new(id: 1234, first_name: 'Dave', last_name: 'Jones')
     missing_user = TSheets::Models::User.new(id: 5678, first_name: 'Darren', last_name: 'Smith')
-    TsheetsAdapter.any_instance.stubs(:find_users).returns([valid_user, missing_user])
+
+    [valid_user, missing_user]
+  end
+end
+
+class TsheetsServiceTest < ActiveSupport::TestCase
+  def setup
+    @service = TsheetsService.new(TsheetsAdapterFake.new)
```

Comment:
You're more keen on dependency injection than mocking. I'm happy either way, but interested to hear your reasoning. :-)

Reply:
Sorry about this, I know there is a school of automated testing that says that everything should be mocked and stubbed. if something brakes then only one test needs to brake as well. I've been following these practices then i was a Java developer and I decided that i don't like this approach.


- usually extensive mocking and stubbing leads to tests that easily brake then you refactor your code. 
- i like to experience all the pain of my code base while I write specs, it helps me to understand which parts of code base is bad and should be improve. With stubbing and mocking you can eliminate all the pain, without improving your code base and hide that pain under the drawer so nobody would notice.


I generally consider, that mocking/stubbing is good for things that are out of your controller -- like external API or services. But usually I resort to vcr gem for that. I've been avoiding all the mocking and stubbing frameworks for more then 4 years already and my testsuits have been very healthy with this ;-)

Again, sorry, that i did this change without consulting/talking with you. But it bothering me too much. :-D

Reply:
Thanks for the explanation, I've not really used dependency injection, but I know it's used pretty extensively in the world outside of ruby, and I certainly don't have anything against it. I generally don't use mocking much either (apart from as you say external services, but also use vcr where possible in that case).

Anyway I'm happy to stick with DI in this project if you prefer that - it would be good to become more familiar with it myself. :-)